### ISSUE TITLE
Inconsistent semantics of close cursor

### ISSUE THREAD

**Issue by lemartin** [doctrine/dbal#2557]
Hello,

I ran into the following problem, using Doctrine DBAL with both MySQL (pdo_mysql) and Oracle (oci8).

Consider the following example, that should run in both MySQL and Oracle.

```php
$statement = $conn->prepare("SELECT ? FROM DUAL");

// first execution
$parameter = 42;
$statement->bindParam(1, $parameter);
$statement->execute();
$statement->fetchAll();

// close cursor
$statement->closeCursor();

// second execution
$parameter = 23;
$statement->bindParam(1, $parameter);
$statement->execute();
$statement->fetchAll();
```

With MySQL it is required (or at least good practice) to close the cursor, freeing any resources held by the result of a prepared statement, enabling the statement to be executed again. Hence in above example, the cursor of the statement is closed after the first execution.

For mysqli `closeCursor` is implemented via `mysqli::free_result`, for pdo_mysql `PDOStatement::closeCursor` is used. In both cases, only any resources held by the result are freed, the prepared statement stays _open_.

> PDOStatement::closeCursor — Closes the cursor, enabling the statement to be executed again.

http://php.net/manual/en/pdostatement.closecursor.php

> mysqli_stmt::free_result -- mysqli_stmt_free_result — Frees stored result memory for the given statement handle

http://php.net/manual/en/mysqli-stmt.free-result.php

With Oracle above example fails when trying to bind `23` during the second execution, the error is `oci_bind_by_name(): 29 is not a valid oci8 statement`. This is easily explained, because for the OCI8 driver, `closeCursor` is implemented using `oci_free_statement`, which closes the prepared statement and not only any resources held by its result.

My problem is now, that `closeCursor` has two very different semantics depending on the concrete driver.

I had a look at the PDO implementation and it seems for databases not having a _cursor_closer_, `closeCursor` is implemented by fetching the remaining rows (see https://github.com/php/php-src/blob/bfb9be9bd4f9629ad501be0720a53c73f7cff285/ext/pdo/pdo_stmt.c#L2060).

The following drivers seem to have special code for closing the cursor (and not the whole statement).

- Firebird https://github.com/php/php-src/blob/ad02235e9f289b76668738d7260cf7684318d647/ext/pdo_firebird/firebird_statement.c#L714
- MySQL https://github.com/php/php-src/blob/6053987bc27e8dede37f437193a5cad448f99bce/ext/pdo_mysql/mysql_statement.c#L896
- Postgre https://github.com/php/php-src/blob/6053987bc27e8dede37f437193a5cad448f99bce/ext/pdo_pgsql/pgsql_statement.c#L699
- SQLite https://github.com/php/php-src/blob/6053987bc27e8dede37f437193a5cad448f99bce/ext/pdo_sqlite/sqlite_statement.c#L348
- SQL Server https://github.com/php/php-src/blob/8176fdcb33f756e6a873330952cd3de3503df1a0/ext/pdo_dblib/dblib_stmt.c#L98

The following drivers do not.

- ODBC https://github.com/php/php-src/blob/6053987bc27e8dede37f437193a5cad448f99bce/ext/pdo_odbc/odbc_stmt.c
- Oracle https://github.com/php/php-src/blob/6053987bc27e8dede37f437193a5cad448f99bce/ext/pdo_oci/oci_statement.c

**Comment by morozov** [comment]
https://github.com/doctrine/dbal/pull/2546 implements freeing only cursor resource and leaving the statement reusable for IBM DB2, Oracle and MS SQL Server.

**Comment by lemartin** [comment]
Yes, that looks exactly like what we need, is there anything I can do to push the pull request. We currently have to use custom drivers classes to circumvent #2557 and #2430 because both pull requests are open.

**Comment by morozov** [comment]
@lemartin as for #2430, it should be fixed by #2536. The latter will make MySQL statement implementation use buffering for subsequent executions and thus prevent the "Commands out of sync" error.

> is there anything I can do to push the pull request

It looks like all maintainers are busy atm, so things go relatively slow.

**Comment by Ocramius** [comment]
Handled in #2546
